torch/csrc/distributed/rpc/tensorpipe_utils.cpp

```diff
@@ -54,6 +179,18 @@ const c10::Stream& getStreamForDevice(
   TORCH_INTERNAL_ASSERT(false, "No stream found for device ", device);
 }
 
+std::array<
+    std::atomic<const TensorpipeDeviceTypeConverter*>,
+    static_cast<size_t>(DeviceType::COMPILE_TIME_MAX_DEVICE_TYPES)>
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+    device_type_converter_registry;
```

Comment:
curious, since this is statically initialized and done in order, why does it still need to be atomic? What type of concurrency are we expecting?

Reply:
It was done this way in the code I copied this from, and this was the original motivation:
https://github.com/pytorch/pytorch/blob/ff15d93b88c0e1e32d2553f272bc654d3b00606f/c10/core/impl/DeviceGuardImplInterface.h#L270-L275

I didn't want to copy all the comments/docstrings so I just mentioned where this code was copied from hoping that readers would then go check there in case of questions. Does that make sense?

Reply:
Yep, that makes sense to me.

> An unsynchronized implementation probably is OK too, but I didn't want  to prove that we never read from device_guard_impl_registry at the same time some registration is occurring.

I still wanna understand our expectations. Would I be correct if I assume in our case, it will be "A unsynchronized implementation is also OK, as TensorPipeAgent is the only owner and all initialization is done in order. The only reason we add atomic here is because this doesn't cost us anything on reads in X86."

Reply:
I must admit I'm not sure. I suspect the main "failure" scenario is as follows: a program starts and loads `libtorch` (i.e., the CPU-only version) because that's what happens by default. Later on, one thread creates a CUDA tensor for the first time, and this triggers `libtorch_cuda` to be loaded, because it's lazy-loaded when needed. At this time, the CUDA DeviceTypeConverter is created and inserted into that array. However at this point there _could_ already be other threads that are trying to read from that array. This is _unlikely_ but technically possible I guess.